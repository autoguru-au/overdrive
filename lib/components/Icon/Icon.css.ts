import { overdriveTokens as vars } from '../../themes/theme.css';
import { makeResponsiveStyle } from '../../utils/responsiveProps.css';

export const size = makeResponsiveStyle(vars.icon.size, (value) => ({
	width: value,
	height: value,
}));
