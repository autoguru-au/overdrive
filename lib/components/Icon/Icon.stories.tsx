import {
	AccountEditIcon,
	AlertCircleIcon,
	CalendarIcon,
	CarIcon,
	CarMultipleIcon,
	CheckIcon,
	CurrencyUsdIcon,
	MagnifyIcon,
	PlusIcon,
	StarIcon,
} from '@autoguru/icons';
import { Meta, StoryFn } from '@storybook/react';
import * as React from 'react';
import { ComponentProps } from 'react';

import { Icon } from '.';

const iconOptions = {
	MagnifyIcon,
	CarIcon,
	CarMultipleIcon,
	CalendarIcon,
	AccountEditIcon,
	AlertCircleIcon,
	CurrencyUsdIcon,
	PlusIcon,
	StarIcon,
	CheckIcon,
	empty: null,
};

export default {
	title: 'Primatives/Icon',
	component: Icon,
	decorators: [],
	argTypes: {
		icon: {
			options: iconOptions,
			control: {
				type: 'select',
			},
		},
		size: {
			options: ['small', 'medium', 'large'],
			control: {
				type: 'select',
			},
		},
	},
} satisfies Meta<typeof Icon>;

const template: StoryFn<typeof Icon> = (args) => <Icon {...args} />;

const standardProps: ComponentProps<typeof Icon> = {
	icon: CalendarIcon,
};
export const Standard = template.bind(standardProps);
Standard.args = standardProps;

const responsiveProps: ComponentProps<typeof Icon> = {
	...standardProps,
	size: ['small', 'medium', 'large'],
};
export const Responsive = template.bind(responsiveProps);
Responsive.args = responsiveProps;

const mediumProps: ComponentProps<typeof Icon> = {
	...standardProps,
	size: 'medium',
};
export const Medium = template.bind(mediumProps);
Medium.args = mediumProps;

const largeProps: ComponentProps<typeof Icon> = {
	...standardProps,
	size: 'large',
};
export const Large = template.bind(largeProps);
Large.args = largeProps;
