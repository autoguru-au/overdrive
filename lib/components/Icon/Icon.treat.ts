import { makeResponsiveStyle } from '../../utils/responsiveProps';

export const size = makeResponsiveStyle(
	({ icon }) => icon.size,
	value => ({
		width: value,
		height: value,
	}),
);
