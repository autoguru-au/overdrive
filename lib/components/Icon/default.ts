export { Icon as default } from './Icon';
