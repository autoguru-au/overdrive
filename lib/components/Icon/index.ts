export { Icon, Props as IIconProps } from './Icon';
export * from '../../icons';
