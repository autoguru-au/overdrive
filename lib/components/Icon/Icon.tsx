import { IconType } from '@autoguru/icons';
import clsx from 'clsx';
import * as React from 'react';
import { ComponentProps, memo, NamedExoticComponent } from 'react';
import { useStyles } from 'react-treat';

import {
	resolveResponsiveStyle,
	ResponsiveProp,
} from '../../utils/responsiveProps';
import { Box } from '../Box';
import * as styleRefs from './Icon.treat';

interface Props extends Pick<ComponentProps<typeof Box>, 'display'> {
	className?: string;
	size?: ResponsiveProp<keyof typeof styleRefs.size>;
	icon: IconType;
}

export const Icon: NamedExoticComponent<Props> = memo(
	({
		className = '',
		icon,
		size = 'small' as ResponsiveProp<keyof typeof styleRefs.size>,
		display,
	}) => {
		const styles = useStyles(styleRefs);

		return (
			<Box
				is="i"
				display={display ?? 'block'}
				className={clsx(
					className,
					resolveResponsiveStyle(size, styles.size),
				)}>
				{icon}
			</Box>
		);
	},
);
