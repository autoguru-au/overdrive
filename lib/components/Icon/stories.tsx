import { CalendarIcon } from '@autoguru/icons';
import * as React from 'react';

import { Icon } from './Icon';

export default {
	title: 'Foundation|Icon',
	component: Icon,
	decorators: [],
};

export const standard = () => (
	<Icon size={['small', 'medium', 'large']} icon={CalendarIcon} />
);
