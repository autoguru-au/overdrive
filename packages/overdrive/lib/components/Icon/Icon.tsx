import type { IconType } from '@autoguru/icons';
import type { FunctionComponent, ReactElement, SVGAttributes } from 'react';
import * as React from 'react';
import { cloneElement } from 'react';

import { resolveResponsiveStyle } from '../../utils/resolveResponsiveProps';
import { ResponsiveProp } from '../../utils/responsiveProps.css';
import type { BoxStyleProps } from '../Box';
import { Box, useBoxStyles } from '../Box';

import * as styles from './Icon.css';

export interface Props {
	display?: Extract<BoxStyleProps['display'], 'block' | 'inlineBlock'>;
	className?: string;
	size?: ResponsiveProp<keyof typeof styles.size | string>;
	icon: IconType | ReactElement<SVGAttributes<SVGElement>, 'svg'>;
}

export const Icon: FunctionComponent<Props> = ({
	className = '',
	icon,
	size = 'small',
	display = 'block',
}) => (
	<Box
		is="i"
		display={display}
		className={[resolveResponsiveStyle(size, styles.size), className]}
		role="presentation">
		{cloneElement(icon, {
			className: useBoxStyles({
				is: 'svg',
				display: 'block',
				width: 'full',
				height: 'full',
			}),
		})}
	</Box>
);
