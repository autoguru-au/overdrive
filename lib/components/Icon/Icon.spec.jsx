import React from 'react';
import { mount, render, shallow } from 'enzyme';
import { Icon } from './Icon';

const TestIcon = (
	<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512">
		<path d="M183.253 353.707L280.96 256l-97.707-97.92 30.08-30.08 128 128-128 128-30.08-30.293z" />
	</svg>
);

describe('<Icon />', () => {
	it('should not throw if SVG icon is passed', () => {
		expect(() => shallow(<Icon icon={TestIcon} />)).not.toThrow();
	});

	it('should pass on className to dom element', () => {
		const icon = mount(<Icon icon={TestIcon} className="icon-class" />);
		expect(icon.hasClass('icon-class')).toBeTruthy();
	});

	it('should match snapshot for same icon', () => {
		const icon = render(<Icon icon={TestIcon} />);
		expect(icon).toMatchSnapshot();
	});

	it('should an i tag for DOM element', () => {
		const icon = shallow(<Icon icon={TestIcon} />);
		expect(icon.type()).toEqual(`i`);
	});

	it('should nest the provided svg inside the i tag', () => {
		const icon = render(<Icon icon={TestIcon} />);
		expect(icon.find('i>svg')).toHaveLength(1);
	});
});
