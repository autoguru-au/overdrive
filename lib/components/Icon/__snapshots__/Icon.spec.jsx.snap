// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon /> should match snapshot for same icon 1`] = `
<i
  class="base display_block small_mobile_base"
>
  <svg
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M183.253 353.707L280.96 256l-97.707-97.92 30.08-30.08 128 128-128 128-30.08-30.293z"
    />
  </svg>
</i>
`;
